### PR TITLE
Refactor module structure etc

### PR DESCRIPTION
**Pull request by tavisrudd**
It was time to split `src/cfn/index` up and clean up the lazy loading logic in `src/cli` (now renamed to `src/main.ts`.

DO NOT MERGE yet.

### CODE REVIEW

File: src/cfn/estimateStackCost.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import {SUCCESS} from '../statusCodes';
```
Comment (jpb) [inline src/cfn/estimateStackCost.ts]: I think this filename should be `TitleCased` (`EstimateStackCost`) given that it is a class.

File: src/cfn/getAllStackEvents.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import * as _ from 'lodash';
+import * as aws from 'aws-sdk'
+
+import def from '../default';
+import eventIsFromSubstack from './eventIsFromSubstack';
+
+export const getSubStacksFromEvents = (events: aws.CloudFormation.StackEvents): Set<string> => {
```
Comment (jpb) [inline src/cfn/getAllStackEvents.ts]: It looks like this is not used outside of the module.
```suggestion
const getSubStacksFromEvents = (events: aws.CloudFormation.StackEvents): Set<string> => {
```

File: src/cfn/listStacks.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          cli.blackBright(lifecyleIcon) + stackName,
+          showTags ? cli.blackBright(prettyFormatTags(stack.Tags)) : ''));
+      if (stack.StackStatus.indexOf('FAILED') > -1 && !_.isEmpty(stack.StackStatusReason)) {
+        console.log('  ', cli.blackBright(stack.StackStatusReason));
+      }
+    }
+  }
+}
+
+export async function listStacksMain(argv: GenericCLIArguments): Promise<number> {
```
Comment (jpb) [inline src/cfn/listStacks.ts]: It seems to me that the CLI-related stuff should not be within the `cfn` directory – what do you think?

File: src/cfn/loadCFNStackPolicy.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import * as _ from 'lodash';
+import {readFromImportLocation, transform} from '../preprocess';
+import maybeSignS3HttpUrl from './maybeSignS3HttpUrl';
+
+
+// TODO is this used?
```
Comment (jpb) [inline src/cfn/loadCFNStackPolicy.ts]: It looks like it is used in `stackArgsToCreateStackInput` for setting the policy during a create (and update?).

File: src/cfn/loadStackArgs.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+import configureAWS from '../configureAWS';
+import getCurrentAWSRegion from '../getCurrentAWSRegion';
+import {logger} from '../logger';
+import {ExtendedCfnDoc, transform} from '../preprocess';
+import * as yaml from '../yaml';
+import {runCommandSet} from './runCommandSet';
+import {StackArgs} from './types';
+import {filter} from '../preprocess/filter';
+
+export function recursivelyMapValues<T extends object>(o: T, f: (a: any) => any): T {
```
Comment (jpb) [inline src/cfn/loadStackArgs.ts]: This isn't used outside of the module.
```suggestion
function recursivelyMapValues<T extends object>(o: T, f: (a: any) => any): T {
```

File: src/cfn/loadStackArgs.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  return _.mapValues(o, function(a: any) {
+    if (typeof a === 'object') {
+      return recursivelyMapValues(a, f);
+    } else {
+      return f(a);
+    }
+  }) as T;
+}
+
+export async function addDefaultNotificationArn(args: StackArgs): Promise<StackArgs> {
```
Comment (jpb) [inline src/cfn/loadStackArgs.ts]: This isn't used outside of the module.

```suggestion
async function addDefaultNotificationArn(args: StackArgs): Promise<StackArgs> {
```

File: src/cfn/maybeSignS3HttpUrl.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import * as aws from 'aws-sdk'
+import * as url from 'url';
+
+export function parseS3HttpUrl(input: string) {
```
Comment (jpb) [inline src/cfn/maybeSignS3HttpUrl.ts]:
```suggestion
function parseS3HttpUrl(input: string) {
```

This isn't used outside of the module.

File: src/cfn/stackArgsToX.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
+import * as aws from 'aws-sdk';
```
Comment (jpb) [inline src/cfn/stackArgsToX.ts]: `s/stackArgsToX/inputHandling/` or something?

File: src/cfn/terminalStackStates.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+// http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-listing-event-history.html
```
Comment (jpb) [inline src/cfn/terminalStackStates.ts]: I think the other status lists could be colocated together, in something like a `statuses` module.

File: src/cfn/terminalStackStates.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+// CREATE_COMPLETE | CREATE_FAILED | CREATE_IN_PROGRESS |
+// DELETE_COMPLETE | DELETE_FAILED | DELETE_IN_PROGRESS |
+// DELETE_SKIPPED | UPDATE_COMPLETexport E | UPDATE_FAILED |
```
Comment (jpb) [inline src/cfn/terminalStackStates.ts]: Lots more of these `exports` in here 😬

File: src/cli-types.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  assumeRoleArn?: string;
+  debug?: boolean;
+  logFullError?: boolean;
+  environment: string;
+  clientRequestToken?: string;
+}
+
+export type GenericCLIArguments = GlobalArguments & yargs.Arguments;
+
+export type ExitCode = number;
```
Comment (jpb) [inline src/cli-types.ts]: Unrelated to the refactor, but could this be limited to the values from `statusCodes`?

File: src/cfn/terminalStackStates.ts (hunk 10 above)
Comment (tavisrudd) [reply src/cfn/terminalStackStates.ts]: There’s a bug in the ts refactoring tool version I started on. I’m going to rebase:rewrite them outfits the history.

File: src/cfn/listStacks.ts (hunk 3 above)
Comment (tavisrudd) [reply src/cfn/listStacks.ts]: I think it should be in there but not littered along side the util code.

File: src/cfn/stackArgsToX.ts (hunk 8 above)
Comment (tavisrudd) [reply src/cfn/stackArgsToX.ts]: Any other `something` suggestions? I couldn't think of any so the current name was my punt.

**Comment by tavisrudd** [comment]
@jpb I've addressed the issues you found.

**Comment by jpb** [comment]
Looks pretty good – the only outstanding thing I would like to see done is to reorganize the `cli-*.ts` files into an actual `cli` module.

File: src/preprocess/visitor.ts [hunk 12]
```diff
@@ -277,6 +277,7 @@ export class Visitor {
       template: node.data.template,
       'var': varName
     });
+    // TODO / BUG? is this a bug: values is unused below
```
Comment (tavisrudd) [inline src/preprocess/visitor.ts]: @jpb what do you think about this?

Comment (tavisrudd) [reply src/preprocess/visitor.ts]: Looks like an value actually used should be equivalent.

**Comment by tavisrudd** [comment]
Going with a merge commit on this one, given the size of the branch.

**Pull request merged**